filecoin-proofs/src/api/seal.rs

```diff
@@ -570,19 +569,50 @@ pub fn seal_commit_phase2<Tree: 'static + MerkleTreeTrait>(
 
     // Verification is cheap when parameters are cached,
     // and it is never correct to return a proof which does not verify.
-    verify_seal::<Tree>(
-        porep_config,
-        comm_r,
-        comm_d,
-        prover_id,
-        sector_id,
-        ticket,
-        seed,
-        &buf,
-    )
-    .context("post-seal verification sanity check failed")?;
-
-    let out = SealCommitOutput { proof: buf };
+    // Non-interactive PoRep is an aggregated proof, hence we also need a different code path for
+    // the verifucation.
+    let out = if porep_config.feature_enabled(ApiFeature::NonInteractivePoRep) {
+        let aggregated = aggregate_seal_commit_proofs::<Tree>(
+            porep_config,
+            &[comm_r],
+            &[seed],
+            &[SealCommitOutput { proof: buf }],
+            groth16::aggregate::AggregateVersion::V2,
+        )?;
+        let inputs = get_seal_inputs::<Tree>(
+            porep_config,
+            comm_r,
+            comm_d,
+            prover_id,
+            sector_id,
+            ticket,
+            seed,
+        )?;
+        let is_valid = verify_aggregate_seal_commit_proofs::<Tree>(
+            porep_config,
+            aggregated.clone(),
+            &[comm_r],
+            &[seed],
+            inputs,
+            groth16::aggregate::AggregateVersion::V2,
+        )
+        .context("post-seal aggregation verification sanity check failed")?;
+        ensure!(is_valid, "post seal aggregation verifies");
+        SealCommitOutput { proof: aggregated }
+    } else {
```

Comment:
If the logic to verify an aggregate proof for NI-PoRep is also being added to `verify_seal`, would something like this work to avoid the duplicate code?
```rust
proof.write(&mut buf)?;

// Non-interactive PoRep is an aggregated proof.
if porep_config.feature_enabled(ApiFeature::NonInteractivePoRep) {
    buf = aggregate_seal_commit_proofs::<Tree>(
        porep_config,
        &[comm_r],
        &[seed],
        &[SealCommitOutput { proof: buf }],
        groth16::aggregate::AggregateVersion::V2,
    )?;
}

let is_valid = verify_seal::<Tree>(..., &buf)
    .context("post-seal verification sanity check failed")?;
ensure!(is_valid, "post seal aggregation verifies");

info!("seal_commit_phase2:finish: {:?}", sector_id);
Ok(SealCommitOutput { proof: buf })
```
But, there may be a reason that I'm not seeing as to why the aggregate proof verification logic for NI-PoRep is needed both here and `verify_seal`.

Reply:
Just adding, by "duplicate logic" I mean the calls to `get_seal_inputs` and `verify_aggregate_seal_commit_proofs` in this `if`-branch.

Reply:
Wow, that was a very good catch! That happens when you work on the code for weeks and in the end you don't see the obvious. I've changed it and pushed it (let's hope I didn't break anything).